clifford/tools/g3c/classify.py

```diff
@@ -0,0 +1,238 @@
+                '{}[{}]'.format(cls.__name__, grade), bases, dict(_grade=grade)
+            )
+            return subcls
+
+    def __call__(cls, *args, **kwargs):
+        """ Overload to prevent `Flat(...)` being used """
+        if not cls.__instantiated:
+            raise TypeError("Must construct as {}[n](...)".format(cls.__name__))
+        return super().__call__(*args, **kwargs)
+
+
+class Blade(metaclass=_GradedTypesMeta):
+    def _repr_skip_members(self):
+        return set()
+
+    def __repr__(self):
+        members = self.__dict__.copy()
+        for name in self._repr_skip_members():
+            members.pop(name)
+        return "{}({})".format(type(self).__name__, ", ".join(
+            "{}={!r}".format(k, v) for k, v in members.items()
+        ))
+
+
+class Direction(Blade):
+    def __init__(self, direction):
+        super().__init__()
+        self.direction = direction
+
+    @property
+    def mv(self):
+        return self.direction^einf
+
+
+class Flat(Blade):
+    def __init__(self, direction, location):
+        super().__init__()
+        self.direction = direction
+        self.location = location
+
+    @property
+    def mv(self):
+        return up(self.location)^self.direction^einf
+
+
+class DualFlat(Blade):
+    def __init__(self, direction, location):
+        super().__init__()
+        self.direction = direction
+        self.location = location
+
+    @property
+    def mv(self):
+        return (up(self.location)^self.direction^einf).dual()
+
+
+class Round(Blade):
+    def __init__(self, direction, location, radius):
+        super().__init__()
+        self.direction = direction
+        self.location = location
+        self.radius = radius
+
+    @property
+    def mv(self):
+        return (up(self.location) + (self.radius*self.radius/2)*einf)^self.direction
+
+
+class Tangent(Round):
+    def __init__(self, direction, location):
+        super().__init__(direction, location, 0)
+
+    def _repr_skip_members(self):
+        return super()._repr_skip_members() | {'radius'}
+
+
+# specializations
+class Point(Tangent[1], override=True):
+    pass
+
+
+class PointFlat(Flat[2], override=True):
+    pass
+
+
+class Line(Flat[3], override=True):
+    pass
+
+
+class Plane(Flat[4], override=True):
+    pass
+
+
+class PointPair(Round[2], override=True):
+    pass
+
+
+class Circle(Round[3], override=True):
+    pass
+
+
+class Sphere(Round[4], override=True):
+    pass
+
+
+class InfinitePoint(Direction[1], override=True):
+    pass
+
+
+def _sqrt(x):
+    """ like cmath.sqrt, but leave real numbers as float """
+    if x < 0:
+        return math.sqrt(-x)*1j
+    else:
+        return math.sqrt(x)
+
+
+def classify(x) -> Blade:
+    """
+    Classify a multivector into a parameterized geomtric description.
+
+    Derived from GA4CS Table 14.1
+
+    Example usage::
+
+        >>> classify(e1)
+        DualFlat[1](location=0, direction=-(1.0^e23))
```

Comment:
This naming seems a little unhelpful to me - is there a better interpretation of these conformal blades?